File: behavioral/php/strategy/Strategy.php
<?php

interface SortStrategy
{
    public function sort(array $items): string;
}


class BubbleSortConcreteStrategy implements SortStrategy
{
    public function sort(array $items): string
    {
        return "Sorting with bubble sort!";
    }
}


class MergeSortConcreteStrategy implements SortStrategy
{
    public function sort(array $items): string
    {
        return "Sorting with merge sort!";
    }
}


class QuickSortConcreteStrategy implements SortStrategy
{
    public function sort(array $items): string
    {
        return "Sorting with quick sort!";
    }
}


class SortContext
{
    private $items;

    private $strategy;


    public function __construct(array $items, SortStrategy $strategy = null)
    {
        $this->items = $items;
        $this->strategy = $strategy;
    }


    public function sortContext()
    {
        if (isset($this->strategy)) {
            return $this->strategy->sort($this->items);
        }

        if (count($this->items) <= 100) {
            return (new BubbleSortConcreteStrategy())->sort($this->items);
        }

        if (count($this->items) <= 10000) {
            return (new MergeSortConcreteStrategy())->sort($this->items);
        }

        return (new QuickSortConcreteStrategy())->sort($this->items);
    }


    public function setStrategy(SortStrategy $strategy)
    {
        $this->strategy = $strategy;
    }
}

File: behavioral/go/strategy/strategy.go
package strategy

type SortStrategy interface {
	Sort() string
}

type BubbleSortConcreteStrategy struct{}

func (b *BubbleSortConcreteStrategy) Sort() string {
	return "Ordenando con bubble sort!"
}

type MergeSortConcreteStrategy struct{}

func (m *MergeSortConcreteStrategy) Sort() string {
	return "Ordenando con merge sort!"
}

type QuickSortConcreteStrategy struct{}

func (q *QuickSortConcreteStrategy) Sort() string {
	return "Ordenando con quick sort!"
}

type SortContext struct {
	items    []uint8
	strategy SortStrategy
}

func NewSortContext(items []uint8, strategy SortStrategy) *SortContext {
	return &SortContext{
		items:    items,
		strategy: strategy,
	}
}

func (sc *SortContext) SortContext() string {
	if sc.strategy != nil {
		return sc.strategy.Sort()
	}

	if len(sc.items) <= 100 {
		return new(BubbleSortConcreteStrategy).Sort()
	}

	if len(sc.items) <= 10000 {
		return new(MergeSortConcreteStrategy).Sort()
	}

	return new(QuickSortConcreteStrategy).Sort()
}

func (sc *SortContext) SetStrategy(strategy SortStrategy) {
	sc.strategy = strategy
}
